include/SDL3/SDL_video_capture.h

```diff
@@ -0,0 +1,353 @@
+ */
+typedef struct SDL_VideoCaptureFrame
+{
+    Uint64 ticks;               /**< ticks when this was read from the driver */
+    int num_planes;             /**< Number of planes */
+    Uint8 *data[3];             /**< Pointer to data of i-th plane */
+    int row_stride[3];          /**< Row stride of i-th plane */
+    void *internal;             /**< Private field */
+} SDL_VideoCaptureFrame;
+
+/**
+ * Open a Video Capture device
+ *
+ * \param device_name device name to open. Can be NULL.
+ * \returns device, or NULL on failure; call
+ *          SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_GetVideoCaptureDeviceName
+ * \sa SDL_GetNumVideoCaptureDevices
+ * \sa SDL_OpenVideoCaptureWithSpec
+ */
+extern DECLSPEC SDL_VideoCaptureDevice *SDLCALL SDL_OpenVideoCapture(const char *device_name);
+
+/**
+ * Set specification
+ *
+ * \param device opened video capture device
+ * \param desired desired video capture spec
+ * \param obtained obtained video capture spec
+ * \param allowed_changes allow changes or not
+ * \returns 0 on success or a negative error code on failure; call
+ *          SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_OpenVideoCapture
+ * \sa SDL_OpenVideoCaptureWithSpec
+ * \sa SDL_GetVideoCaptureSpec
+ */
+extern DECLSPEC int SDLCALL SDL_SetVideoCaptureSpec(SDL_VideoCaptureDevice *device,
+                                                    const SDL_VideoCaptureSpec *desired,
+                                                    SDL_VideoCaptureSpec *obtained,
+                                                    int allowed_changes);
+
+/**
+ * Open a Video Capture device and set specification
+ *
+ * \param device_name device name to open. Can be NULL.
+ * \param desired desired video capture spec
+ * \param obtained obtained video capture spec
+ * \param allowed_changes allow changes or not
+ * \returns device, or NULL on failure; call
+ *          SDL_GetError() for more information.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_OpenVideoCapture
+ * \sa SDL_SetVideoCaptureSpec
+ * \sa SDL_GetVideoCaptureSpec
+ */
+extern DECLSPEC SDL_VideoCaptureDevice *SDLCALL SDL_OpenVideoCaptureWithSpec(const char *device_name,
+                                                                              const SDL_VideoCaptureSpec *desired,
+                                                                              SDL_VideoCaptureSpec *obtained,
+                                                                              int allowed_changes);
+/**
+ * Get device name
+ *
+ * \param index device index between 0 and num -1
+ * \returns device name, shouldn't be freed
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_GetNumVideoCaptureDevices
+ */
+extern DECLSPEC const char * SDLCALL SDL_GetVideoCaptureDeviceName(int index);
```

Comment:
This should switch to using device IDs instead of by index, the way we've done with sensors, joysticks, and audio devices. I can easily see someone hotplugging a webcam, so the same rationale applies here.

You would add SDL_GetVideoCaptureDevices() that returned an array of SDL_VideoCaptureDeviceID, and then you'd have a function to get the name of the device by ID, and then you'd open the device by ID instead of by name. This will avoid race conditions where the device list changes unexpectedly while being enumerated and opened. Even if you don't support hotplug now, that's something we'll want in the future (along with the associated added/removed events)

Reply:
there was this patch before: 
https://github.com/libsdl-org/SDL/commit/425033200ac0392dcb1d64b3ea5c99be4ea1cb02
should I re-add this  ?

Reply:
No, using pointers is great, you just want to identify the devices by ID before they're opened.
e.g. from the joystick API:
```c
extern DECLSPEC SDL_JoystickID *SDLCALL SDL_GetJoysticks(int *count);
extern DECLSPEC const char *SDLCALL SDL_GetJoystickInstanceName(SDL_JoystickID instance_id);
extern DECLSPEC SDL_Joystick *SDLCALL SDL_OpenJoystick(SDL_JoystickID instance_id);
```

Reply:
ok, I'll check this next week !